File: src/main/java/blog/typed/javadsl/ChatRoom.java
package blog.typed.javadsl;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import akka.actor.typed.ActorRef;
import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.Actor;

public abstract class ChatRoom {
  private ChatRoom() {
  }

  static interface Command {}
  public static final class GetSession implements Command {
    public final String screenName;
    public final ActorRef<SessionEvent> replyTo;
    public GetSession(String screenName, ActorRef<SessionEvent> replyTo) {
      this.screenName = screenName;
      this.replyTo = replyTo;
    }
  }

  private static final class PostSessionMessage implements Command {
    public final String screenName;
    public final String message;
    public PostSessionMessage(String screenName, String message) {
      this.screenName = screenName;
      this.message = message;
    }
  }

  static interface SessionEvent {}
  public static final class SessionGranted implements SessionEvent {
    public final ActorRef<PostMessage> handle;
    public SessionGranted(ActorRef<PostMessage> handle) {
      this.handle = handle;
    }
  }
  public static final class SessionDenied implements SessionEvent {
    public final String reason;
    public SessionDenied(String reason) {
      this.reason = reason;
    }
  }
  public static final class MessagePosted implements SessionEvent {
    public final String screenName;
    public final String message;
    public MessagePosted(String screenName, String message) {
      this.screenName = screenName;
      this.message = message;
    }
  }

  public static final class PostMessage {
    public final String message;
    public PostMessage(String message) {
      this.message = message;
    }
  }

  public static Behavior<Command> behavior() {
    return chatRoom(new ArrayList<ActorRef<SessionEvent>>());
  }

  private static Behavior<Command> chatRoom(List<ActorRef<SessionEvent>> sessions) {
    return Actor.immutable(Command.class)
      .onMessage(GetSession.class, (ctx, getSession) -> {
        ActorRef<PostMessage> wrapper = ctx.spawnAdapter(p ->
          new PostSessionMessage(getSession.screenName, p.message));
        getSession.replyTo.tell(new SessionGranted(wrapper));
        List<ActorRef<SessionEvent>> newSessions =
          new ArrayList<ActorRef<SessionEvent>>(sessions);
        newSessions.add(getSession.replyTo);
        return chatRoom(newSessions);
      })
      .onMessage(PostSessionMessage.class, (ctx, post) -> {
        MessagePosted mp = new MessagePosted(post.screenName, post.message);
        sessions.forEach(s -> s.tell(mp));
        return Actor.same();
      })
      .build();
  }



}

File: src/main/scala/blog/typed/scaladsl/ChatRoom.scala
package blog.typed.scaladsl

import scala.io.StdIn

import akka.actor.typed.ActorSystem
import akka.actor.typed.Behavior
import akka.actor.typed.scaladsl.Actor
import akka.actor.typed.ActorRef
import akka.actor.typed.Terminated

object ChatRoom {
  sealed trait Command
  final case class GetSession(screenName: String, replyTo: ActorRef[SessionEvent])
    extends Command
  private final case class PostSessionMessage(screenName: String, message: String)
    extends Command

  sealed trait SessionEvent
  final case class SessionGranted(handle: ActorRef[PostMessage]) extends SessionEvent
  final case class SessionDenied(reason: String) extends SessionEvent
  final case class MessagePosted(screenName: String, message: String) extends SessionEvent

  final case class PostMessage(message: String)

  val behavior: Behavior[Command] =
    chatRoom(List.empty)

  private def chatRoom(sessions: List[ActorRef[SessionEvent]]): Behavior[Command] =
    Actor.immutable[Command] { (ctx, msg) ⇒
      msg match {
        case GetSession(screenName, client) ⇒
          val wrapper = ctx.spawnAdapter {
            p: PostMessage ⇒ PostSessionMessage(screenName, p.message)
          }
          client ! SessionGranted(wrapper)
          chatRoom(client :: sessions)
        case PostSessionMessage(screenName, message) ⇒
          val mp = MessagePosted(screenName, message)
          sessions foreach (_ ! mp)
          Actor.same
      }
    }
}

object ChatRoomApp {

  import ChatRoom._

  val gabbler =
    Actor.immutable[SessionEvent] { (_, msg) ⇒
      msg match {
        case SessionDenied(reason) ⇒
          println(s"cannot start chat room session: $reason")
          Actor.stopped
        case SessionGranted(handle) ⇒
          handle ! PostMessage("Hello World!")
          Actor.same
        case MessagePosted(screenName, message) ⇒
          println(s"message has been posted by '$screenName': $message")
          Actor.stopped
      }
    }

  def main(args: Array[String]): Unit = {
    val root: Behavior[akka.NotUsed] =
      Actor.deferred { ctx ⇒
        val chatRoom = ctx.spawn(ChatRoom.behavior, "chatroom")
        val gabblerRef = ctx.spawn(gabbler, "gabbler")
        chatRoom ! GetSession("ol’ Gabbler", gabblerRef)

        Actor.empty
      }

    val system = ActorSystem(root, "ChatRoomDemo")
    try {
      // Exit the system after ENTER is pressed
      StdIn.readLine()
    } finally {
      system.terminate()
    }
  }
}
